% !TEX encoding = UTF-8 Unicode

\documentclass[10pt]{article}
\usepackage[top=20mm, bottom=20mm, left=20mm, right=20mm]{geometry}
\geometry{a4paper}
\geometry{portrait}
\usepackage[utf8]{inputenc}	% Comment out for xelatex; must use with pdflatex
\usepackage[T1]{fontenc}
\usepackage[parfill]{parskip} % Activate to begin paragraphs with an empty line rather than an indent
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{amssymb}
%\usepackage{fontspec}
\usepackage{epstopdf}
\usepackage{float}
\usepackage[hang]{footmisc}
\usepackage{endnotes}
\usepackage{listingsutf8}
%\restylefloat{table}
\usepackage{xcolor}
\usepackage{colortbl}
%\usepackage{booktabs}
\usepackage{tabu}
\usepackage{caption}
\usepackage{textcomp}
\usepackage{textgreek}
\usepackage{subcaption}
\usepackage{hyperref}
\usepackage[normalem]{ulem}
\usepackage{authblk}
\usepackage[numbers]{natbib}
\usepackage{xr}
\externaldocument{supplementary}

\bibliographystyle{plos2015}


\makeatletter
\def\blx@maxline{77}
\makeatother

\captionsetup[figure]{labelfont=sc}
\captionsetup[table]{labelfont=sc}
\captionsetup[tabu]{labelfont=sc}

\DeclareCaptionType{MPEquation}[][List of equations]
\captionsetup[MPEquation]{labelformat=empty}

\DeclareCaptionType{MPListing}[][List of listings]
\captionsetup[MPListing]{labelformat=empty}

\usepackage{fancyhdr}
\setlength{\headheight}{16pt}
\pagestyle{fancyplain}
\fancyhf{}
%\fancypagestyle{plain}{}
\cfoot[\thepage]{\thepage}

\usepackage{stringenc}
\usepackage{pdfescape}


% Helper for debugging pdflatex errors induced by invalid UTF8 characters:
% output the actual character code causing the issue.

\makeatletter
\renewcommand*{\UTFviii@defined}[1]{%
\ifx#1\relax
\begingroup
% Remove prefix "\u8:"
\def\x##1:{}%
% Extract Unicode char from command name
% (utf8.def does not support surrogates)
\edef\x{\expandafter\x\string#1}%
\StringEncodingConvert\x\x{utf8}{utf16be}% convert to UTF-16BE
% Hexadecimal representation
\EdefEscapeHex\x\x
% Enhanced error message
\PackageError{inputenc}{Unicode\space char\space \string#1\space
(U+\x)\MessageBreak
not\space set\space up\space
for\space use\space with\space LaTeX}\@eha
\endgroup
\else\expandafter
#1%
\fi
}
\makeatother

\inputencoding{utf8}

% Make multi-paragraph footnotes align nicely on the left
\setlength\footnotemargin{10pt}

% Do not display an automatic Notes heading for endnotes, as they are contained
% by a Manuscripts section with a heading that has possibly been edited by the user
\def\enoteheading{}

% Apply footnote and endnote numbering schemes (decimal, lowercase roman etc.)
\renewcommand{\thefootnote}{\arabic{footnote}}
\renewcommand{\theendnote}{\arabic{endnote}}
% Adjust endnotes to use normal-sized numbering and text
\renewcommand{\enotesize}{\normalsize}
\renewcommand\enoteformat{%
  \raggedright
  \leftskip=1.8em
  \makebox[0pt][r]{\theenmark. \rule{0pt}{\dimexpr\ht\strutbox+\baselineskip}}%
}


\begin{document}
\title{Nested Stochastic Block Models Applied to the Analysis of Single Cell Data}
\author[1,2]{Leonardo Morelli}
\author[1,3]{Valentina Giansanti}
\author[1]{Davide Cittaro}
\affil[1]{Center for Omics Sciences, IRCCS San Raffaele Institute, Milan, Italy}
\affil[2]{Universit\`{a} Vita-Salute San Raffaele, Milan, Italy} 
\affil[3]{Department of Informatics, Systems and Communication, University of Milano-Bicocca, Milan, Italy}
\maketitle

\section*{Abstract}

Single cell profiling has been proven to be a powerful tool in molecular biology to understand the complex behaviours of heterogeneous system. The definition of the  properties of single cells is the primary endpoint of such analysis, nevertheless cells are typically clustered to underpin the common determinants that can be used to describe functional properties of the cell mixture under investigation. Several approaches have been proposed to identify cell clusters; while this is matter of active research, one popular approach is based on community detection in neighbourhood graphs by optimisation of modularity. In this paper we propose an alternative and principled solution to this problem, based on Stochastic Block Models. We show that our approach is not only able to correctly identify cell groups, but also adds the possibility to perform inference on cell identities \textcolor{red}{SOMETHING MORE? Damn, I always struggle with abstracts}.


\section*{Background}

Transcriptome analysis at single cell level by RNA sequencing (scRNA-seq) is a technology growing in popularity and applications \cite{svensson_2018}. It has been applied to study the biology of complex tissues \cite{guo_2018, ventotormo_2018}, tumor dynamics \cite{rozenblattrosen_2020, tirosh_2016, patel_2014, neftel_2019}, development \cite{rosenberg_2018, wagner_2018} and to describe whole organisms \cite{plass_2018, regev_2017}.

A key step in the analysis of scRNA-seq data and, more in general, of single cell data, is the identification of cell populations, that is groups of cells sharing similar properties. Several approaches have been proposed to achieve this task, based on well established clustering techniques \cite{wang_2017, lin_2017}, consensus clustering \cite{huh_2020, kiselev_2017, Ranjan_2021} and deep learning \cite{li_2020}; many more have been recently reviewed \cite{krzak_2019, kiselev_2019} and benchmarked \cite{du_2018}. As the popularity of single cell analysis frameworks \texttt{Seurat} \cite{butler_2018} and \texttt{scanpy} \cite{wolf_2018} raised, methods based instead on graph partitioning became the \emph{de facto} standards. Such methods require the construction of a cell neighbourhood graph (\emph{e.g.} by \emph{k }Nearest Neighbours, \emph{k}NN, or shared Nearest Neighbors, \emph{s}NN). Encoding cell-to-cell similarities into graphs has practical advantages beyond clustering, as many algorithms for graph analysis can be applied and interpreted in a biological way. A notable example is the analysis of cell trajectories which can be derived from the analysis of Markov processes traversing the NN graph \cite{Setty_2019, Lange_2020}. In another context, computation of RNA moments in scRNA velocity is also based on the NN graph structure \cite{Bergen_2020}. Arguably, the biggest utility of NN structure is the possibility to identify cell groups by partitioning the graph into communities; this is typically achieved using the Louvain method \cite{blondel_2008}, a fast algorithm for optimisation of graph modularity. While fast, this method does not guarantee that small communities in large networks are well defined. To overcome its limits, a more recent approach, the Leiden algorithm \cite{traag_2019}, has been implemented and it has been quickly adopted in the analysis of single cell data, for example by \texttt{scanpy} \cite{wolf_2018} and \texttt{PhenoGraph} \cite{levine_2015}. In addition to Newman's modularity \cite{newman_2004}, other definitions currently used in single cell analysis make use of a resolution parameter \cite{traag_2011, reichardt_2006}. In lay terms, resolution works as a threshold on the density within communities: lowering the resolution results in less and sparser communities and \emph{vice versa}. Identification of an appropriate resolution has been recognised as a major issue \cite{lhnemann_2020}, also because it requires the definition of a mathematical property (clusters) over biological entities (the cell groups), with little formal description of the latter. In addition, the larger the dataset, the harder is to identify small cell groups, as a consequence of the well-known resolution limit \cite{fortunato_2007}. Moreover, it has been demonstrated that random networks can have modularity \cite{guimer_2004} and its optimisation is incapable of separating actual structure from those arising simply of statistical fluctuations of the null model. Lastly, it is a common error to assume that the resolution parameter reflects a hierarchical structure of the communities in the graph when, in general, this is not rigorously true. Additional solutions to cell group identification from NN graphs have been proposed, introducing resampling techniques \cite{baran_2019, Tang_Sackton_2020} or clique analysis \cite{xu_2015}. It has been proposed that high resolution clustering, \emph{e.g.} obtained with Leiden or Louvain methods, can be refined in agglomerative way using machine learning techniques \cite{miao_2020}.

An alternative solution to community detection is the Stochastic Block Model, a generative model for graphs organized into communities \cite{holland_1983}. In this scenario, identification of cell groups requires the estimation of the proper parameters underlying the observed NN graph. According to the microcanonical formulation \cite{peixoto_2017}, the parameters are partitions and the matrix of edge counts between them. Under this model, nodes belonging to the same group have the same probability to be connected together. It is possible to include node degree among the model parameters \cite{karrer_2011}, to account for heterogeneity of degree distribution of real-world graphs. A Bayesian approach to infer parameters has been developed \cite{peixoto_2013} and implemented in the \texttt{graph-tool} python library (\href{https://graph-tool.skewed.de}{https:/\slash graph-tool.skewed.de}). There, a generative model of network $\boldsymbol A$ has a probability $P(\boldsymbol A|\boldsymbol\theta, \boldsymbol b)$ where \textbf{$\boldsymbol\theta$} is the set of parameters and \textbf{\emph{$\boldsymbol b$}} is the set of partitions. The likelihood of the network being generated by a given partition can be measured by the posterior probability


\begin{equation}
P(\boldsymbol b | \boldsymbol A) = \frac{P(\boldsymbol A|\boldsymbol\theta, \boldsymbol b)P(\boldsymbol\theta, \boldsymbol b)}{P(\boldsymbol A)}
\label{Equation_posterior}
\end{equation}

and inference is performed by maximising the posterior probability. The numerator in Eq. \ref{Equation_posterior} can be rewritten exponentiating the description length

\begin{equation}
\Sigma = -\ln P(\boldsymbol A|\boldsymbol\theta, \boldsymbol b) - \ln P(\boldsymbol\theta, \boldsymbol b)
\label{Equation_entropy}
\end{equation}

so that inference is performed by minimizing the information required to describe the data (Occam's razor); \texttt{graph-tool} is able to efficiently do this by a Markov Chain Monte Carlo approach \cite{peixoto_2014}. SBM itself may fail to identify small groups in large graphs, hence hierarchical formulation has been proposed \cite{peixoto_2014_h}. Under this model, communities are agglomerated at a higher level in a block multigraph, also modelled using SBM. This process is repeated recursively until a graph with a single block is reached, creating a Nested Stochastic Block Model (nSBM).

In this work we propose nSBM for the analysis of single cell data, in particular scRNA-seq data. This approach identifies cell groups in a statistical robust way and, moreover, it is able to determine the likelihood of the grouping, thus allowing model selection. In addition, it is possible to measure the confidence of assignment to groups, a measure that can be exploited in various analysis tasks.

We developed \texttt{schist} (\href{https://github.com/dawe/schist}{https:/\slash github.com\slash dawe\slash schist}), a python library compatible with \texttt{scanpy}, to facilitate the adoption of stochastic block models in single-cell analysis.

\section*{Results}

\subsection*{Overview of \texttt{schist}}

\texttt{schist} is a convenient wrapper to the \texttt{graph-tool} python library, written in python and designed to be used with \texttt{scanpy}. The most prominent function is \texttt{schist.inference.nested\_model()} which takes a \texttt{AnnData} object as input and fits a nested Stochastic Block Model on the \emph{k}NN graph built with \texttt{scanpy} functions (\emph{e.g.} \texttt{scanpy.tools.neighbors()}). When launched with default parameters, \texttt{schist} fits a model which maximises the posterior probability of having a set of cell groups (or blocks) given a graph. \texttt{schist} then annotates cells in the data object with all the groups found at each level of a hierarchy. Given the large size of the NN graph in real-world experiments, it is possible that a single solution represents local minima of the fitting process. In addition, it is possible that multiple solutions are equally acceptable to represent the graph partitioning and a better description is given by the consensus over such solutions \cite{Peixoto_2021_consensus}. To overcome these issues, \texttt{schist} fits multiple instances in parallel and returns the inferred consensus model, alongside the marginal probabilities for each cell to belong to a specific group (\emph{cell marginals}). In addition to the default nested Stochastic Block Model, \texttt{schist} implements the possibility to fit a Bayesian model able to find statistically significant assortative modules \cite{Zhang_Peixoto_2020}, eliminating the need to set a resolution parameter as required in standard community detection by maximization of modularity.

Once a model has been fitted, \texttt{schist} allows to evaluate the difference in entropy given by assigning every cell to every possible group. This step allows to estimate the strength of association of a cell to a specific group (\emph{cell affinity}). We show that cell affinity may have practical application in common tasks in single cell analysis, such as label transfer.

\subsection*{Absence of information is correctly handled by nSBM}

One of the most relevant difference between SBM and other methods to cluster single cells is that it relies on robust statistical modelling. In this sense, the number of groups identified strictly mirrors the amount of information contained in the data. An important consequence is that absence of information can be properly handled. To show this property we performed a simple experiment on a randomised \emph{k}NN graph. We collected data for 3k PBMC (available as preprocessed data in \texttt{scanpy}) and shuffled the edges of the prebuilt \emph{k}NN graph, this to keep the general graph properties unchanged. We tested that the degree distribution does not change after randomisation (Kolmogorov-Smirnov $D=0.0733$, $p=0.703$). We found that the default strategy, based on maximisation of modularity, identifies 24 cell groups, whereas nSBM does not identify any cell group. 

\begin{figure}[H]
\centering
\includegraphics[keepaspectratio,width=0.6\textwidth,height=\textheight]{FIgure_Random.png}
\caption[]{nSBM results for randomised data. (A) UMAP embedding of ~3k PBMC after standard processing with Leiden approach (left) or nSBM (right). Cell grouping is consistent for both the approaches (Adjusted Rand Index $ARI=0.869$). (B) UMAP embedding of the same data after randomisation of \emph{k}NN graph edges. In this case nSBM does not return any cell grouping, while optimisation of modularity finds up to 24 different cell groups}\label{FigureRandom}
\end{figure}

This experiment is a deliberate extreme case. The quality of grouping proposed by a standard approach can be disputed in many ways, and the UMAP embedding indeed reflects the absence of any information. Nevertheless, real-world data may include an unknown amount of random noise. Hence, It is important to identify cell groups that are not artefacts arising from processing and that do reflect the information contained in the dataset. 

\subsection*{\texttt{schist} correctly identifies cell populations}

To benchmark \texttt{schist}, we tested it on scRNA-seq mixology data \cite{Tian_2019}, a dataset explicitly developed to benchmark single cell analysis tools without the need to simulate data. In particular, we used the mixture of 5 cell lines profiled with Chromium 10x platform. At a first evaluation of the UMAP embedding, all lines appear well separated. Only the lung cancer line H1975 shows a considerable degree of heterogeneity and appears to be split into two cell groups (Fig. \ref{FigureTian}A). Using default parameters, \texttt{schist} is able to identify correct cell groups ($ARI=0.829$), with a further split in H2228 cell line (Fig. \ref{FigureTian}B), whereas Leiden method clusters the dataset into 10 groups ($ARI=0.549$, Fig. \ref{FigureTian}C). The nSBM solution correctly identifies H1975 groups as a single entity at level 1 of the hierarchy. We then sought to check if an independent agglomerative method, SCCAF \cite{miao_2020}, was able to recover cell line groupings starting from both partition schemes. Given the ground truth, the cell lines, SCCAF is able to assess the maximal accuracy that can be achieved in the dataset (0.992). When trained with this target accuracy, SCCAF precisely reconstructs the original cell line annotations starting from \texttt{schist} partitions with high accuracy (Fig. \ref{FigureTian}D). When Leiden partitions are set as input, SCCAF merges H2228 and HCC827 cells into a single cluster and keeps H1975 cells split into two groups (Fig. \ref{FigureTian}E), highlighting potential limitations of this approach. In order to perform a fair and complete experiment, we also tested SCCAF with the partitions scheme at the deepest level of the nSBM hierarchy (38 groups, Fig. \ref{Tian_NSBM0}A), for which we observed the inability to perform a significant cell grouping (Fig. \ref{Tian_NSBM0}B). 

\begin{figure}[H]
\centering
\includegraphics[keepaspectratio,width=0.6\textwidth,height=\textheight]{FigureTian.png}
\caption[]{Analysis of a mixture of 5 known cell lines. (A) UMAP embedding colored and labeled by cell line identity. H1975 cells (orange) show two distinct groups reflecting an internal heterogeneity. (B) UMAP embedding showing cells colored by level 1 of the hierarchy proposed by the nested Stochastic Block Model. (C) UMAP embedding showing cells colored according to the Leiden method at resolution $\gamma = 1$. (D) UMAP embedding colored by the classification made by SCCAF when partitions in (B) are used. (E) UMAP embedding colored by the classification made by SCCAF when partitions in (C) are used.}\label{FigureTian}
\end{figure}

In all, these data support the suitability of \texttt{schist}, hence of Nested Stochastic Block Models, for cell group identification in single cell studies. 

\subsection*{Hierarchy modelling is accurate in defining biological properties}

When grouping is performed by optimization of modularity, there is often the implicit assumption that the resolution parameter reflects a hierarchical structure of the graph, that is communities are consistently grouped at lower resolutions. Not only this assumption is wrong, but it may also lead to spurious groupings in real experiments, whereas nSBM inherently encodes hierarchies by merging communities in a tree. The improper use of resolution parameter may lead to two types of errors: grouping of cells that are in fact distinct and creating an inconsistent hierarchy. 

To show this we took advantage of public spatial RNA dataset of a coronal section of murine brain tissue profiled with 10X Visium H\&E technology \cite{Gracia_Villacampa_2020}, as provided by the recently introduced package \texttt{SquidPy} \cite{Palla_Theis_2021}. We chose to stick to the given tissue annotation by the package authors. At default resolution, Leiden clustering resolves the tissue structure, as does the first level of the nSBM hierarchy (Fig. \ref{Figure_Visium}). When resolution is decreased (\emph{e.g.} $\gamma = 0.5$), the dentate gyrus is incorrectly merged to the hippocampus, whereas nSBM correctly identifies the pyramidal layer. 

\begin{figure}[H]
\centering
\includegraphics[keepaspectratio,width=\textwidth,height=0.55\textheight]{Figure_Visium_Spatial.png}
\caption[]{Analysis of spatial transcriptomics of a coronal section of mouse brain by Visium H\&E. In the first panel, original tissue annotation is given. Tissues are well defined at default resolution for the standard approach. When resolution is decreased to $\gamma = 0.5$, cells from the dentate gyrus are merged to the hippocampus. When nSBM is used, the structure of the pyramidal is maintained at different levels of the hierarchy}\label{Figure_Visium}
\end{figure}

In another context, we tested the effect on the interpretation of the hierarchy varying the resolution parameter. We analysed data for hematopoietic differentiation \cite{paul_2015}, previously used to benchmark the consistency of cell grouping with differentiation trajectories by graph abstraction \cite{wolf_2019} (Fig. \ref{Figure_Hemato_Supp}A). Data show three major branchings (Erythroids, Neutrophils and Monocytes) stemming from the progenitor cells, mostly recapitulated by level 2 of the hierarchy computed by \texttt{schist} (Fig. \ref{Figure_Hemato_Lineage}). Not only the hierarchic model recapitulates the branching trajectories, also the cell groups appear to be consistent with the estimated pseudotime. Conversely, the Leiden method at default resolution identified 24 groups. By lowering the $\gamma$ parameter we observed cell groups that merge and split at different resolutions disrupting the hierarchy (Fig. \ref{Figure_Paul15_Leiden_r}). 


\begin{figure}[H]
\centering
\includegraphics[keepaspectratio,width=0.6\textwidth,height=\textheight]{Figure_Hemato_Lineage.png}
\caption[]{Analysis of hematopoietic differentiation. Each panel presents a low dimensional embedding of single cells next to a radial tree representation of the nSBM hierarchy. Cells are colored according to groupings at level 2 of the hierarchy, group 0 marks the most primitive population (A). In subsequent panels, cells are colored using a signature of Erythroid lineage (B), Monocytes (C) or Neutrophils (D).}\label{Figure_Hemato_Lineage}
\end{figure}

In all, these data show that the common intuition that $\gamma$ parameter acts as a thresholding factor over a hierarchy is wrong. Not only the hierarchy is not conserved, but also very different cell types may be mixed in spurious clusters. By using nSBM, \texttt{schist} is able to represent hierarchical relations in appropriate way. Moreover, the hierarchy appears to be more robust in aggregating different cell types at coarser scales.

\subsection*{Cell marginals can be used to assess the data quality}

By computing the consensus among multiple models, \texttt{schist} returns the marginal probability for each cell to belong to a specific cluster at each level of the hierarchy. Ideally, all cells should always be assigned with $p=1$ to a cluster. When the uncertainty is maximal, cells are assigned to clusters randomly with $p=1/B_i$, where $B_i$ is the number of groups for the $i$-th level in the hierarchy. We sought to check if these probabilities could be interpreted in terms of data quality. 

We devised a simple metric, \emph{cell stability}, that is defined by the fraction of levels for which the marginal probability is higher than $1-1/B_i$. To do so, we only consider levels with at least two groups, hence excluding the root of the tree. We tested this metric on four datasets from \cite{mereu_2020} with different quality levels (iCELL8, MARS-seq, 10XV3 and Quartszeq-2) (Fig. \ref{Figure_Stability}). By taking a summary metric, \emph{e.g.} the mean $\overline{S}$ or the fraction of cells with $S>0.5$, we observed that it correlates with the data quality (Table \ref{table_stability}).

\begin{table}[h!]
\centering
 \begin{tabular}{|| l c c ||}
 \hline
 \textbf{Dataset} & \textbf{$\overline{S}$} & \textbf{$S > .5$} \\ [0.5ex] 

 \hline\hline
 iCELL8 \cite{mereu_2020} & 0.368 & 0.312  \\
 \hline
 MARS-seq \cite{mereu_2020} & 0.579 & 0.536  \\
 \hline
 Chromium 10x \cite{mereu_2020} & 0.716 & 0.728  \\
 \hline
 Quartz-seq2 \cite{mereu_2020} & 0.705 & 0.739 \\
 \hline
\end{tabular}
\caption{Cell stability as indicator of data quality. Table shows summary metrics derived from the Cell Stability calculated for various datasets. $\overline{S}$ is the average Cell Stability over all cells, $S>.5$ indicates the fraction of cells with Cell Stability higher than 0.5.}
\label{table_stability}
\end{table}

These data suggest that measures of uncertainty of cell clustering can be useful for general quality control assessment. In addition to this, we foresee they could be used to isolate cells with specific patterns.

%\subsection*{Cell marginals have little impact on detection of marker genes}

%We sought to investigate if cell marginals can be used as continuous covariates when testing for differentially expressed genes among cell groups. To this end, we took advantage of the \texttt{Splatter} package \cite{Zappia_Oshlack_2017} to simulate 6 cell populations at different levels of noise (biological coefficient of variation, $BCV$, ranging from 0.2 to 2) and probabilities to have differentially expressed genes within a group ($p(DE)$, ranging from 0.1 to 1). For each cell group we built two Generalized Linear Models, one with dummy variables recapitulating the standard \emph{t}-test that is performed by \texttt{scanpy} (\emph{i.e.} $\sim1 + group$) and one using the marginal (\emph{i.e.} $\sim1 + group\_marginal$). In general, we did not observe substantial changes between the two models (Fig. \ref{Figure_DE}). Using marginals slightly increases the amount of positive tests (at $FDR<0.01$) and the sensitivity of the method, especially when the data are noisy, but at the cost of the reduction of recall due to an increased rate of false positives. In all, the utility of cell marginals to this task remains dubious.

\subsection*{Cell affinities can be used for label transfer}

The modelling approach we adopted allows the estimation of the information required to describe a graph given any partitioning scheme, not limited to the solution given by the model itself. Differences in entropy can be used to perform model selection, that is we can choose which model better describes the data. We sought to exploit this property to address the task of annotating cells according to a reference sample. To this end we analyzed datasets from \cite{mereu_2020}, which includes mixtures of human PBMC and HEK293T cells profiled with various technologies. We chose cells profiled with 10X V3 platform as reference dataset and performed annotation on cells profiled with Quartz-seq2 or MARS-seq. These are at the extremes of the capability to distinguish cell types, so they provide good benchmark configurations for this task. 

After preprocessing raw data according to the parameters given in \cite{mereu_2020}, we integrated each dataset with 10XV3 into a unified representation using Harmony \cite{Korsunsky_2019}, and computed the \emph{k}NN graph. In each merged dataset, we retained cell type annotations for 10X cells, while we assigned a "Unknown" label to all cells derived from the other technology (\emph{i.e.} MARS-seq or Quartz-seq2). We then calculated the cell affinity matrix, that is we computed the difference in entropy that can be observed by assigning each cell to each annotation cluster, this being either one of the original cell types or "Unknown". Once the matrix has been computed, each cell from the query data is assigned to the group with the highest likelihood. The rationale behind this approach is that if cells belong to the same annotation group, then more information is required to describe the graph if they were annotated as different cell types; hence, cells from the query datasets should retain their "Unknown" label if and only if there is not enough evidence to associate them to another group. We compared the accuracy of the  outcome to \emph{k}NN classification, given by the closest entry in the \emph{k}NN graph, and to \texttt{ingest}, a tool included in \texttt{scanpy}  based on \emph{k}NN classification of UMAP embeddings. Analysis of a well defined dataset, such as Quartz-seq2, reveals that the three approaches are equally good in classifying unknown cells (Fig. \ref{Figure_Label_Transfer}, central column), with accuracies ranging from .870 to .927. When data are noisy, instead, \emph{k}NN-based methods show low accuracy and a tendency to assign the most represented cell group (HEK293T) to the unlabeled cells. This misannotation is particular evident for \texttt{ingest}, in which only CD4 T cells and HEK cells are transferred, resulting in the lowest accuracy (0.243). Conversely, \texttt{schist} is able to assign correct labels with higher accuracy (0.641). Moreover, \emph{k}NN methods assign a label to each cell, whereas \texttt{schist} does not relabel cells if there are no sufficient evidence (\emph{e.g.} the "Unknown" state is the most likely). Interestingly, we found that for the largest part of cells without assigned label, the second choice by affinity ranking was indeed the appropriate one (Fig. \ref{hist_label_second}).

\begin{figure}[H]
\centering
\includegraphics[keepaspectratio,width=0.9\textwidth,height=\textheight]{Figure_Label_Transfer.png}
\caption[]{Label transfer using SBM. The first line reports UMAP embeddings for datasets profiled with Chromium 10X V3, Quartz-seq2 and  MARS-seq, each annotated by known cell types. Quartz-seq2 and MARS-seq were reannotated using \emph{k}NN method, \texttt{sc.tl.ingest() } or \texttt{schist}. The accuracy of each label transfer task is reported above the corresponding UMAP.} \label{Figure_Label_Transfer}
\end{figure}

\subsection*{Choice of an optimal hierarchy level}

The Nested Stochastic Block Model, hence \texttt{schist}, fits a hierarchical model of communities into a graph. When it comes to analysis of single cell data, it means that the cells are best described by the hierarchy itself and that cells \emph{can} be grouped consistently at each level of the tree. In addition, the size of groups at the deepest level scales as $O(N/\log{N})$ \cite{peixoto_2014_h}, where $N$ is the number of cells. Given the current throughput in single cell experiments ($\sim$10k cells), the number of groups becomes intractable. For this reason, in most of single cell experiments, it is preferable to identify an optimal level of the hierarchy that best resembles the cell properties at the scale they can be validated. 

A possible strategy is based on Random Matrix Theory, as suggested by the authors of the SC3 package  \cite{kiselev_2017}, for which a suitable number of clusters, $\hat{k}$, is determined by the number of eigenvalues of the $\boldsymbol Z^\top\boldsymbol Z$ matrix (where $\boldsymbol Z$ is the normalized count matrix) significantly different at $p<.001$ from the appropriate Tracy-Widom distribution. According to this strategy, the optimal level $i^k$ is the one that minimizes the number of partitions and $\hat{k}$:

\begin{equation}
i^k = \underset{x}{\mathrm{argmin}} |B_x - \hat{k}|
\label{Equation_RMT}
\end{equation}

where $B_x$ is the number of non empty partitions at level $x$.

An alternative strategy is to evaluate the behaviour of modularity at different hierarchy levels. While nSBM does not optimize the graph modularity $Q$, we observed that this tends to be maximal for the level better describing known cell populations, so the optimal level $i^Q$ is

\begin{equation}
i^Q = \underset{x}{\mathrm{argmax}} |Q_x|
\label{Equation_MODmax}
\end{equation}

Where $Q_x$ is modularity at level $x$. We collected values arising from both the approaches for some datasets used in this work (Table \ref{table_optimal})

\begin{table}[h!]
\centering
 \begin{tabular}{|| l c c c c c c c ||}
 \hline
 \textbf{Dataset} & \textbf{Cells} & \textbf{$D$}  & \textbf{$\hat{k}$} & \textbf{$i^k$} & \textbf{$B_k$} & \textbf{$i^Q$} & \textbf{$B_Q$} \\ [0.5ex] 

 \hline\hline
 sc-mixology \cite{Tian_2019} & 860 & 5  & 21 & 1 & 6 & 1 & 6 \\ 
\hline
 Chromium 10x \cite{mereu_2020} & 1523 & 8  & 43 & 0 & 58 & 1 & 13 \\
 \hline
 Quartz-seq2 \cite{mereu_2020} & 1266 & 8  & 37 & 0 & 62 & 1 & 12 \\
 \hline
 MARS-seq \cite{mereu_2020} & 1401 & 9  & 9 & 1 & 16 & 1 & 16 \\
 \hline
 iCELL8 \cite{mereu_2020} & 1830 & 9  & 20 & 1 & 21 & 2 & 6 \\
 \hline
 Mouse brain \cite{Gracia_Villacampa_2020} & 2688 & 15  & 8 & 2 & 8 & 1 & 23 \\
 \hline
 Planaria \cite{plass_2018} & 21612 & $51^\ast$ & 34 & 2 & 22 & 3 & 10 \\
 \hline
\end{tabular}
\caption{Selection of the optimal level in the nSBM hierarchy. For each dataset we report the number of groups $D$ that were given by the authors. The optimal level selection should recover a number of groups in the order of magnitude of $D$. Value of $D$ in Planaria dataset is derived from manual curation of Louvain clustering. $\hat{k}$:  number of groups according to RMT, $i^k$: level selected according to RMT criterion, $B_k$: number of partitions at level $i^k$, $i^Q$: level at which modularity is maximal, $B_Q$: number of groups at level $i^Q$}
\label{table_optimal}
\end{table}

As expected, the larger the network, the higher the optimal level. For relatively small datasets (\emph{i.e.} less than 10k cells), the first level of the hierarchy contains a number of groups in line with how many observable populations are. 
%\subsection*{PPBM?}

%\textcolor{red}{TO BE DONE. Maybe this is the right place to introduce the choice of the appropriate hierarchy level}


\subsection*{Analysis of runtimes}

Minimisation of the nSBM is a process that requires a large amount of computational resources. While the underlying \texttt{graph-tool} library is efficient in exploring the solution space using a multiflip MCMC sampling strategy, the number of required iterations before convergence is considerable and the running time scales linearly with the number of edges. Moreover, to collect a consensus partition, we minimize multiple models (default: 100) that need to be averaged. To give a reference, we report runtimes for some example datasets in Table \ref{Table_runtime} on a commodity hardware (Intel i7@2.8 GHz, 32 GB RAM). Compared to Leiden approach, nSBM requires at best $\sim6\times$ times more, and $\sim30\times$  at worst. A reasonably fast alternative to the nSBM is the Planted Partition Block Model (PPBM), for which we also report runtimes. The PPBM \cite{Zhang_Peixoto_2020} is able to find statistically significant assortative modules and eliminates the resolution parameter; differently from nSBM, PPBM is not hierarchic.

\begin{table}[h!]
\centering
 \begin{tabular}{|| l c c c c c ||}
%  \begin{tabular}{|| m{7em} | m{5em} | m{7em} | m{7em} | m{5em} ||}
 \hline
 \textbf{Dataset} & \textbf{Cells} & \textbf{Edges} & \textbf{Leiden} & \textbf{PPBM} & \textbf{nSBM} \\ [0.5ex] 
 \hline\hline
 sc-mixology \cite{Tian_2019} & 860 & 9186 & 00:06 & 00:13 & 00:36 \\ 
 \hline
 Quartz-seq2 \cite{mereu_2020} & 1266 & 14603 & 00:10 & 00:19 & 00:45 \\
 \hline
 MARS-seq \cite{mereu_2020} & 1401 & 21756 & 00:20 & 00:34 & 02:14 \\
 \hline
 iCELL8 \cite{mereu_2020} & 1830 & 30636 & 00:23 & 00:40 & 03:02 \\
 \hline
 Chromium 10x \cite{mereu_2020} & 1523 & 21447 & 00:14 & 00:26 & 01:07 \\ 
% \hline
% Islets \cite{Segerstolpe_2016} & 2108 & 32023 & 00:20 & 00:42 & 03:13 \\
 \hline
 Hematopoiesis \cite{paul_2015} & 2730 & 15444 & 00:37 & 01:27 & 05:52\\ 
 \hline
 Mouse Cortex \cite{Zeisel_2015} & 3005 & 54460 & 00:59 & 00:53 & 07:32 \\
 \hline
 Endocrinogenesis \cite{Bastidas_Ponce_2019} & 3696 & 74670 & 01:15 & 01:29 & 10:56 \\
 \hline
 Baron Pancreas \cite{Baron_2016} & 8569 & 294480 & 03:51 & 07:35 & 1:33:40 \\
 \hline
 Airzani Liver \cite{Aizarani_2019} & 10368 & 354440 & 04:13 & 09:47 & 1:42:23 \\
 \hline
 Planaria \cite{plass_2018} & 21612 & 173667 & 05:41 & 13:52 & 1:20:40 \\
 \hline
\end{tabular}
\caption{Time required to run different partitioning strategies implemented in \texttt{schist} on various datasets. All approaches fit 100 models. Number of nodes and edges refer to the structure of the \emph{k}NN graph as built by \texttt{scanpy}. Times are expressed in mm:ss. }
\label{Table_runtime}
\end{table}

\section*{Conclusions}

Identification of cells sharing similar properties in single cell experiments is of paramount importance. A large number of approaches have been described, although the standardisation of analysis pipelines converged to methods that are based on modularity optimisation. We tackled the biological problem using a different approach, nSBM, which has several advantages over existing techniques. The most important advantage is the hierarchical definition of cell groups which eliminates the choice of an arbitrary threshold on clustering resolution. In addition, we showed that the hierarchy itself could have a biological interpretation, implying that the hierarchical model is a valid representation of the cell ensemble. 

The Bayesian formulation of Stochastic Block Models provides the possibility to perform inference on a graph for any partition configuration, thus allowing reliable model selection using an interpretable measure, entropy. We exploited this property to perform label transfer with high accuracy and with the possibility to discard cells with unreliable assignments. In all, \texttt{schist} facilitates the adoption of nSBM by the bioinformatics community and exposes a robust framework to perform tasks that go beyond the principled identification of cell clusters. 

The major drawback of adopting this strategy is the substantial increase of runtimes. As observed, model minimisation may be hundred times slower than the extremely fast Leiden approach. It should be noted that \texttt{schist} initializes multiple models that are treated by multiple concurrent processes. \texttt{graph-tool} itself supports CPU-level parallelization for some of its tasks. These optimizations are well suited for clustered computing infrastructure. Further development, possibly including GPU-level parallelization, is surely required to accomodate the large size of datasets that are being produced.

\section*{Materials and Methods}

Unless differently stated, all the analysis were produced using \texttt{scanpy} v1.7.1 \cite{wolf_2018} and \texttt{schist} v0.7.6 and the corresponding dependencies. All models were initialized 100 times, herein including Leiden partitioning for which we also calculated the consensus partition.

\subsection*{Analysis of Random data}

Data were retrieved in \texttt{scanpy} environment using \texttt{sc.datasets.pbmc3k\_processed()} function. The random \emph{k}NN graph was obtained shuffling the node labels of each edge. UMAP embedding was recomputed after randomisation using the shuffled graph.

\subsection*{Analysis of cell mixtures}

Data and metadata for five cell mixture profiled by Chromium 10x were downloaded from the sc-mixology repository (\href{https://github.com/LuyiTian/sc_mixology}{https:/\slash github.com\slash LuyiTian\slash sc\_mixology}). Cells with less than 200 genes were excluded, as genes detected in less than 3 cells. Cells with less than 5\% of mitochondrial genes were retained for subsequent analysis. Data were normalised and log-transformed; number of genes and percentage of mitochondrial genes were regressed out. \emph{k}NN graph was built with default parameters (50 components and 15 nearest neighbors). Data were assessed by SCCAF using cell line annotation. Mean cross-validated accuracy was set as target for all the models.

\subsection*{Analysis of Visium H\&E data}
Data were retrieved using \texttt{squidpy.datasets.visium\_hne\_adata()} built-in function, without further processing. Leiden clustering was performed using \texttt{schist.inference.leiden()} function, allowing for 100 initializations, with resolutions $\gamma = 1$ and $\gamma = 0.5$. 

\subsection*{Analysis of hematopoietic differentiation}

Data were retrieved using \texttt{scanpy}'s built-in functions and were processed as in \cite{wolf_2019}, except for \emph{k}NN graph built using 30 principal components, 30 neighbours and diffmap as embedding. Gene signatures were calculated using the following gene lists

\begin{itemize}
\item Erythroids: Gata1, Klf1, Epor, Gypa, Hba-a2, Hba-a1, Spi1
\item Neutrophils, Elane, Cebpe, Ctsg, Mpo, Gfi1
\item Monocytes, Irf8, Csf1r, Ctsg, Mpo
\end{itemize}


\subsection*{Processing of PBMC data from various platforms}
Count matrices were downloaded from GEO using the following accession numbers: GSE133535 (Chromium 10Xv3), GSE133543 (Quartz-seq2), GSE133542 (MARS-seq) and GSE133541 (iCELL8). Data were processed according to the methods in the original paper \cite{mereu_2020}. Briefly, cells with less than 10,000 total number of reads as well as the cells having less than 65\% of the reads mapped to their reference genome were discarded. Cells in the 95th percentile of the number of genes/cell and those having less than 25\% mitochondrial gene content were included in the downstream analyses. Genes that were expressed in less than five cells were removed. Data were normalized and log-transformed, highly variable genes were detected at minimal dispersion equal to 0.5. Neighbourhood graph was built using 30 principal components and 20 neighbours. 


\subsection*{Label transfer}

Processed data for MARS-seq or Quart-seq2 platforms were merged to data for 10X V3. Merged data were then processed using Harmony \cite{Korsunsky_2019} by the \texttt{scanpy.external.pp.harmony\_integrate()} function with default parameters. Cells not belonging to the 10X data were assigned an "Unknown" label. We calculated cell affinity to each annotation label using \texttt{schist.tl.calculated\_affinity()} function. We assigned the most affine annotation only to "Unknown" cells. For \emph{k}NN-based procedure, we built a \emph{k}NN graph on the merged data using \texttt{pynndescent } library on the 10XV3 subset of cells in the merged data, then we assigned "Unknown" cells to the closest entry in the graph. Assignment by \texttt{scanpy.tl.ingest()} was performed using default parameters.



\section*{Acknowledgements}

We would like to thank Tiago de Paula Peixoto (Central European University, ISI Foundation) and Giovanni Petri (ISI Foundation) for the discussions and the precious hints. We also would like to thank all people at COSR, in particular Giovanni Tonon and Paolo Provero.
This work has been supported by Accelerator Award: A26815 entitled:  "Single-cell cancer evolution in the clinic" funded through a partnership between Cancer Research UK and Fondazione AIRC




\bibliography{references}


\end{document}
